### ISSUE TITLE
Ensure portability of top-level module names (case-sensitivity)?

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#4678]
On case-sensitive file systems, the following is legal Agda:
```agda
import A
import a
```
assuming files `A.agda` and `a.agda` are in the project root.

However, this might lead to non-portable code when moving to a case-insensitive file system.

I would advocate that Agda should disallow source files that only differ on case in the project root, as this might lead to portability problems.

**Comment by nad** [comment]
Are there other potential pitfalls related to file names?

**Comment by andreasabel** [comment]
> Are there other potential pitfalls related to file names?

I suppose so.  On Mac I can have a `module A\B where` in a file `A\B.agda`, but this may not fly on Windows (don't actually know).

**Comment by nad** [comment]
I think we should aim for maximum portability:
* Only printable 7-bit ASCII characters.
* Not `/`, `\`, or `.`, and no lower-case letters. In fact, for simplicity, let's only allow `A` to `Z`.
* At least one file system does not support directories, so hierarchical module names with more than one component have to go.
* At most eight characters in the module name.
* Let's replace `.agda` with `.AGD`.
* Come to think of it, we should support EBCDIC as well. The overlap between ASCII and EBCDIC is small, but there are nineteen control characters that have the same codes when these two character encodings are used. Let us use these instead of `A` to `Z`.

**Comment by andreasabel** [comment]
https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations

Probably too boring:

POSIX "Fully portable filenames"[22] 
- allowed: A–Z a–z 0–9 . _ - 
- reserved: / null 
- hyphen must not be first character

But it would make sense to disallow common reserved characters:

E.g. FAT family: 0x00-0x1F 0x7F " * / : < > ? \ | + , . ; = [ ] (in some environments also: ! @; DOS 1/2 did not allow 0xE5 as first character)
NTFS: 0x00-0x1F 0x7F " * / : < > ? \ |

**Comment by jespercockx** [comment]
How do other programming languages handle this (e.g. Haskell, Coq, Idris, ...)? This does not seem like a problem where we should try to reinvent the wheel.

**Comment by nad** [comment]
If I recall correctly the Haskell 98 report does not specify how modules should be stored: there is no obligation to store one module per file, it is fine to, say, store all the code in a database.

**Comment by andreasabel** [comment]
@nad And a serious contribution to the discussion?

**Comment by UlfNorell** [comment]
GHC checks that the capitalization of the file name matches the module name. On case insensitive file systems, the file name you read isn't necessarily the file name that is stored on disk, so you can get confusing errors like this:

(On MacOS)
```haskell
-- Foo.hs
module Foo where
foo x = x
```
```haskell
-- Bar.hs
module Bar where
import FOO
bar = foo "foo"
```
```
λ :l Bar
FOO.hs:2:8: error:
    File name does not match module name:
    Saw: ‘Foo’
    Expected: ‘FOO’
  |
2 | module Foo where
  |        ^^^
Failed, no modules loaded.
```

**Comment by andreasabel** [comment]
@UlfNorell I reported a respective bug to GHC a while ago: https://gitlab.haskell.org/ghc/ghc/issues/16507

**Comment by UlfNorell** [comment]
It doesn't look like they're taking it very seriously...

**Comment by andreasabel** [comment]
No, this is cosmetics.  We already check this, as far as I know.
However, I don't think GHC allows to import modules that only differ in capitalization.  We should also disallow it, in my opinion.

**Comment by UlfNorell** [comment]
How do you propose to implement this? What if the case-mismatched imports are in different files? Surely, we shouldn't traverse the file system from the project root looking for potential clashes.

My feeling is that it's hard to come up with and implement a sensible restriction that also provides some real guarantees.

**Comment by UlfNorell** [comment]
Also, for the record
```haskell
-- Foo.hs
module Foo where
```
```haskell
-- FOO.hs
module FOO where
```
```haskell
-- Main.hs
module Main where
import Foo
import FOO
```
```
$ ghci Main
GHCi, version 8.8.1: https://www.haskell.org/ghc/  :? for help
[1 of 3] Compiling FOO              ( FOO.hs, interpreted )
[2 of 3] Compiling Foo              ( Foo.hs, interpreted )
[3 of 3] Compiling Main             ( Main.hs, interpreted )
Ok, three modules loaded.
```

**Comment by andreasabel** [comment]
I tried the setup with include paths `src1` and `src2` and files `src1/a.agda` and `src2/A.agda`.  In my main file I import both:
```agda
import a
import A
```
I am a bit surprised that Agda does report on Mac the error:
```
Ambiguous module name. The module name a could refer to any of the
following files:
 .../Issue4678/src1/a.agda
 .../Issue4678/src2/A.agda
when scope checking the declaration
  import a
```
On a case-sensitive fs this should pass at the moment.  Even one a case-preserving fs like Mac this *could* be made to work (which I do not advocate).  On a case-insensitive fs this could never work.
Thus, it seems that we do not have to do much on top of the existing logic to get portability.  Currently, this is how module names resolve to files:
https://github.com/agda/agda/blob/014312d0181796ffc26557177bebe59bd8584c4c/src/full/Agda/Interaction/FindFile.hs#L143-L167
There is a case-sensitive file finder for Windows:
https://github.com/agda/agda/blob/014312d0181796ffc26557177bebe59bd8584c4c/src/full/Agda/Utils/FileName.hs#L107-L122
So, if I understand correctly, the logic in `findFile''` tries to handle my case with `src1/a` and `src2/A` and maybe does so on Windows, but not on Mac.
I suppose we could swap `doesFileExistCaseSensitive` for a `doesFileExistCaseInsensitive` and get a portable behavior.

**Comment by andreasabel** [comment]
Proposal how to modify:
https://github.com/agda/agda/blob/014312d0181796ffc26557177bebe59bd8584c4c/src/full/Agda/Interaction/FindFile.hs#L154-L157
* Turn `files` into a list of `Glob` patterns replacing each char by an alternative between its lower and upper case version. https://hackage.haskell.org/package/Glob  Make sure to not include the paths or file extensions into the globbing, only the parts coming from the module name.
* Construct `existingFiles` by a call to `globDir` https://hackage.haskell.org/package/Glob-0.10.0/docs/System-FilePath-Glob.html#v:globDir

**Comment by UlfNorell** [comment]
Sounds reasonable I think, although I'm a little hesitant to pull in more questionably maintained dependencies over this issue...

**Comment by andreasabel** [comment]
It would probably imply that I have to volunteer for maintaining Glob. It's currently up for grabs.

**Comment by andreasabel** [comment]
I have been bitten by this again today.
I think we should not allow file-modules that just differ on case, for the sake of our users who might not be aware of wide-spread case-insensitive file systems (like the default on macOS).

**Comment by nad** [comment]
> Are there other potential pitfalls related to file names?

We could perhaps use [`System.FilePath.Posix.isValid`](https://hackage.haskell.org/package/filepath-1.4.100.0/docs/System-FilePath-Posix.html#v:isValid) and [`System.FilePath.Windows.isValid`](https://hackage.haskell.org/package/filepath-1.4.100.0/docs/System-FilePath-Windows.html#v:isValid), and reject file names that are not accepted by both tests.
